### PR TITLE
Verification for Python Topology

### PR DESCRIPTION
**Pull request by omus**
It is quite possible I missed something with the verification code as I am new to Storm. Let me know if I did and I'll be happy to correct any issues.

### CODE REVIEW

**Comment by amontalenti** [comment]
Did a quick glance at this, it looks pretty awesome! :+1:

I'm assigning the PR review to @dan-blanchard, since I think he'll need to think through whether we want the DSL in pystorm or not.

(Also, I'm travelling today, so won't be able to review until later this week.)

Thanks for the work on this, @omus!

**Comment by amontalenti** [comment]
also cc @cabiad @kbourgoin @msukmanowsky @dfdeshom on this Topology DSL validation work.

**Comment by omus** [comment]
Thanks @amontalenti. I didn't noticed the Python 3 issues with metaclass until I made the pull request. I'll try to sort those out while I wait for feedback.

File: streamparse/storm/bolt.py [hunk 1]
```diff
@@ -9,14 +9,76 @@
...
+    def resolve_dependencies(self, specifications):
+        """Modifies the Bolt's sources to be references to other Specification
+        objects. Also ensures that group_on fields are provided by the
+        source Specifications.
+
+        :param specifications: all of the specification objects for this
+                               topology.
+        :type specifications: dict
+        """
+        for i in range(len(self.sources)):
```
Comment (dan-blanchard) [inline streamparse/storm/bolt.py]: Minor nitpick:

```python
for i, source in enumerate(self.sources):
```

would be a bit cleaner.

**Comment by dan-blanchard** [comment]
Thanks for the contribution! Overall, this looks really nice.  The only substantial thing I think we'll want to change is what @amontalenti hinted at with his comment:

> I'm assigning the PR review to @dan-blanchard, since I think he'll need to think through whether we want the DSL in pystorm or not.

I think we should keep the DSL out of the `streamparse.storm` sub-package, which will soon be split off as `pystorm`.  That way people who aren't using streamparse to submit the topologies and such don't have this extra method (`spec`) that they don't understand what it's for.  The question then becomes how do we want to deal with `Component.spec`, if we don't want to have that included in `pystorm`.

I see two possible approaches:

1.  Forget about `Component.spec` entirely, and just have people instantiate the `Spec` class directly.
2.  Once the current  `streamparse.storm` package is split off as `pystorm`, create new `streamparse.storm.component.Component`, `streamparse.storm.bolt.Bolt`, and `streamparse.storm.spout.Spout` classes that inherits from the real `pystorm` ones and just add the `spec` method.

Personally, I prefer the first approach, as it's simpler and won't lead to potential confusion down the line having a subclasses that are named the same thing as their parents (except for the namespace).

File: streamparse/storm/bolt.py (hunk 1 above)
Comment (omus) [inline streamparse/storm/bolt.py]: I definitely agree.

**Comment by omus** [comment]
I also am in favor of using the `Spec` class over `Component.spec`.

**Comment by dan-blanchard** [comment]
I got rid of `Component.spec`, refactored things slightly (to move all of the DSL-related stuff into the `dsl` package), and merged in master.  I'm going to merge this into master now, since it looks good to go.

Thanks for your excellent contribution!

**Pull request merged**